cdap-data-fabric/src/main/java/co/cask/cdap/data2/transaction/queue/hbase/HBaseQueueDebugger.java

```diff
@@ -0,0 +1,280 @@
+                            TransactionExecutorFactory txExecutorFactory) {
+    this.queueAdmin = queueAdmin;
+    this.queueClientFactory = queueClientFactory;
+    this.zkClientService = zkClientService;
+    this.txExecutorFactory = txExecutorFactory;
+  }
+
+  @Override
+  protected void startUp() throws Exception {
+    zkClientService.startAndWait();
+  }
+
+  @Override
+  protected void shutDown() throws Exception {
+    zkClientService.stopAndWait();
+  }
+
+  /**
+   * Only works for {@link co.cask.cdap.data2.transaction.queue.hbase.ShardedHBaseQueueStrategy}.
+   */
+  private void scanQueue(final QueueName queueName) throws Exception {
+    HBaseConsumerStateStore stateStore = queueAdmin.getConsumerStateStore(queueName);
+
+    TransactionExecutor txExecutor = Transactions.createTransactionExecutor(txExecutorFactory, stateStore);
+    txExecutor.execute(new TransactionExecutor.Procedure<HBaseConsumerStateStore>() {
+      @Override
+      public void apply(HBaseConsumerStateStore stateStore) throws Exception {
+        Multimap<Long, QueueBarrier> barriers = stateStore.getAllBarriers();
+        System.out.printf("Got %d barriers", barriers.size()).println();
+
+        QueueStatistics stats = new QueueStatistics();
+        int currentSection = 1;
+
+        for (Long groupId : barriers.keySet()) {
+          System.out.printf("Scanning barriers for group %s", groupId).println();
+          Collection<QueueBarrier> groupBarriers = barriers.get(groupId);
+          PeekingIterator<QueueBarrier> barrierIterator = Iterators.peekingIterator(groupBarriers.iterator());
+          while (barrierIterator.hasNext()) {
+            QueueBarrier start = barrierIterator.next();
+            QueueBarrier end = barrierIterator.hasNext() ? barrierIterator.peek() : null;
+
+            System.out.printf("Scanning section %d/%d...", currentSection, barriers.size()).println();
+            scanQueue(stateStore.getTransaction(), queueName, start, end, stats);
+            System.out.printf("Current results: %s", stats.getReport()).println();
+            currentSection++;
+          }
+          System.out.println("Scanning complete");
+        }
+
+        System.out.printf("Total results: %s", stats.getReport()).println();
+      }
+    }, stateStore);
+  }
+
+  private void scanQueue(Transaction tx, QueueName queueName, QueueBarrier start,
+                         @Nullable QueueBarrier end, QueueStatistics outStats) throws IOException {
+
+    byte[] queueRowPrefix = QueueEntryRow.getQueueRowPrefix(queueName);
+
+    ConsumerGroupConfig consumerConfig = start.getGroupConfig();
+    System.out.printf("Got consumer group config: %s", consumerConfig).println();
+
+    if (!queueClientFactory.tableExists(queueName)) {
+      System.out.printf("Found no flow queue table for %s", queueName).println();
+      return;
+    }
+
+    HBaseQueueAdmin admin = queueClientFactory.ensureTableExists(queueName);
```

Comment:
I notice you check tableExists earlier, however it's dangerous to have code like this because if someone just move the code around a bit, it will start creating new table. See my comment above to just rely on the `TableNotFoundException` from HBase to tell if the table exists or not.